q2_diversity_lib/tests/test_beta.py

```diff
@@ -266,14 +275,21 @@ def test_accepted_types_have_consistent_behavior(self):
                                             self.expected[id1, id2])
 
     def test_does_it_run_through_framework(self):
-        unweighted_unifrac_thru_framework = self.plugin.actions[
-                    'unweighted_unifrac']
-        table_as_artifact = Artifact.import_data(
-                    'FeatureTable[Frequency]', self.table_as_BIOMV210Format)
-        tree_as_artifact = Artifact.import_data(
-                    'Phylogeny[Rooted]', self.tree_as_NewickFormat)
-        unweighted_unifrac_thru_framework(table_as_artifact,
-                                          tree_as_artifact)
+        self.unweighted_unifrac_thru_framework(self.table_as_artifact,
+                                               self.tree_as_artifact)
+        # If we get here, then it ran without error
+        self.assertTrue(True)
+
+    @mock.patch("q2_diversity_lib._util.psutil.Process")
+    def test_cpu_request_through_framework(self, mock_process):
+        mock_process = psutil.Process()
+        mock_process.cpu_affinity = mock.MagicMock(return_value=[0, 1, 2])
+        self.unweighted_unifrac_thru_framework(self.table_as_artifact,
+                                               self.tree_as_artifact,
+                                               threads=2)
+        self.unweighted_unifrac_thru_framework(self.table_as_artifact,
```

Comment:
Good tests! Do you have a similar test, for checking if `n_jobs` == `auto` works as expected?

Reply:
Hoo boy am I glad you asked. New commit with breaking test incoming. Something very messy is happening. I suspect it's related to the fact that everything is being unpacked into positional arguments in our return statement, but I'm really not sure. 

Some notes:
`jaccard` fails with `TypeError: jaccard() got multiple values for argument 'n_jobs'`
`unifrac` methods don't trigger test failures, however...
when jaccard fails, a stderr message is exposed that comes from [from unifrac](https://github.com/biocore/unifrac/blob/8ab40f0c09b5975450d6f28d2ad7cc2806ba6f06/sucpp/api.cpp#L261) that makes me think it's getting a memory address or something instead of the integer we're trying to deliver: 
```
------------------------------------ Captured stderr call -------------------------------------
More threads were requested than stripes. Using 342021232 threads.
```
The number in the error message is a different large integer (positive or negative) every time. 
These `stderror` lines appear only when there is a test that uses `auto`. Integer arguments are clean, which doesn't seem to jive well with my current hypothesis on why this is happening. :man_shrugging:

Reply:
I think you might've found a bug in the somewhere in the framework or test harness! First off, the main problem is in the test data itself:

```python
        self.jaccard_thru_framework(self.table_as_artifact,
                                    self.tree_as_artifact,
                                    n_jobs=2)
```

Note, you are passing in a tree as the second param, even though this method doesn't take a tree. That appears to be you second `n_jobs` (which is what I think might be a bug in the framework.

While playing with this, my next thought was, "well, is this somehow related to the decorator signature munging?" - I removed the decorators and tested, the answer appears to be "no". Put another way, I don't see how this could be related to your decorator utils (yet). 

Applying the following patch to 8dfdb0b has a 100% passing test suite:

```patch
diff --git a/q2_diversity_lib/tests/test_util.py b/q2_diversity_lib/tests/test_util.py
index 13cacab..8db7a63 100644
--- a/q2_diversity_lib/tests/test_util.py
+++ b/q2_diversity_lib/tests/test_util.py
@@ -181,10 +181,8 @@ class ValidateRequestedCPUsTests(TestPluginBase):
                                                self.tree_as_artifact,
                                                threads='auto')
         self.jaccard_thru_framework(self.table_as_artifact,
-                                    self.tree_as_artifact,
                                     n_jobs=2)
-        # self.jaccard_thru_framework(self.table_as_artifact,
-        #                             self.tree_as_artifact,
-        #                             n_jobs='auto')
+        self.jaccard_thru_framework(self.table_as_artifact,
+                                    n_jobs='auto')
         # If we get here, then it ran without error
         self.assertTrue(True)
```

Reply:
Here's a MWE:

```python
import qiime2 
from qiime2.plugins import demux 
dmx = qiime2.Artifact.load('../data/moving-pictures/demux.qza')
demux.actions.summarize(dmx, dmx, dmx, dmx, n=dmx) 
```

Reply:
To clarify my thinking RE a bug in the framework. I know that this "got multiple values" business is just python doing what python does, but I am thinking there is a guard or two that we might be missing, although I'm not sure.

Reply:
Totally agreed on this - seems like we should be doing a better job of checking argument counts somewhere. I'll go poking around with this in a little bit. 

For now, I've figured out what's causing the `stderr` messages from `unifrac`, I'm just not sure why they display the numbers that they do. In every case where that stderror output appears, we were working with a small (3-sample) table, which `unifrac` can handle using a maximum of two threads. Using a larger table removes the issue, as does requesting a smaller number of threads. The 'auto' parameter had nothing to do with it, beyond that it was requesting 3 threads, one more than the max. 

I'm wondering: 
- Why is unifrac reporting using 46530796 or -132493215 threads?
- Should we be "catching" this error somehow, and passing a warning along to users? Unifrac seems to be handling the problem gracefully (and reporting it), but the strange thread counts reported concern me.
- Is this an issue I should raise on `unifrac`, at least to start a conversation?